[WIP]feat: OPC UA Role Permissions

AsishGanesh1 opened the pull request:
Attach well known roles to session

 - [X] Add 8 well known role details to session instance
         UA_Server_getSessionAttribute API can be made use of for accessing the role data
 - [X] Remove redundant API's (like hasAccessToNode, hasAccessToMethod)
 - [X] Remove read user AccessLevel/Executable/RolePermissions related methods
         To avoid read outside the context of a session

 **TODO:**
 - [ ] Add CMake flag for usage of role based permission
 - [ ] Add test application
 - [ ] Address remaining review comments
 - [ ] Rebase to current master head

<asish@eclatron.com>

opcua-tsn-team-kalycito committed:
[TASK] Implement check database

 - Register callback while client tries to connect
   to server using username

Change-Id: Ic85121614c84ab70f93f4572b0024679ea30627d

opcua-tsn-team-kalycito committed:
[TASK] Implement add and remove role functionality

 - This commit adds the default roles in address space and
   handles the add and remove role functionality

Change-Id: Icecd4e197d9b1071390b207828a9b7c723483050

opcua-tsn-team-kalycito committed:
[TASK] Implementation of role permission access for milestone 3

 - This commit handles different roles based on session

Signed-off-by: Keerthivasan.AS<keerthivasan.as@kalycito.com>
Change-Id: I2e588fef4019ddb06c8928200158b30362c16cb9

opcua-tsn-team-kalycito committed:
[TASK] Implementation to add required optional method nodes for milestone 4

  - Adds the AddIdentity and RemoveIdentity method nodes
    in the address space
  - Fixes the userExecutable permission rights
  - Fixes the logger function used instead of printf()
  - Add IdentityCriteriaMappingType in the Identity nodes created
  - AddIdentity and RemoveIdentity method calls handling
  - Fix memory leaks and remove value change in AddIdentity() method

Change-Id: Ic6663f724e2a5756bb2ac8b6b154d67c4f11aa3e

opcua-tsn-team-kalycito committed:
[TASK] Implementation to add Role permission for milestone 5

  - This commit adds the role and user role node along with
    the updated permission for a single node
  - This commit fixes the segementation fault for wrong
    user name and password
  - This commit adds client code for server validation

Change-Id: I27a318782100a94335cec56d6c3ff8fd76b73591

opcua-tsn-team-kalycito committed:
[TASK] Implementation to test Role permission for milestone 6

  - This commit adds test cases for role based implementation
  - Fixes the logging mechanism
  - Added unit test cases

Change-Id: I55ff9810115d8c82d0b0b5cf07c9d3149b9fcfd9

opcua-tsn-team-kalycito committed:
feat(server): Implementation of roles permission (opcua part 18)

-> renamed test_role_permissions.c to check_role_permissions.c
-> removed "Experimental" message when enabling A&C.
Signed-off-by: Keerthivasan Alagarsamy Senthilkumaran <keerthivasan.as@kalycito.com>
Change-Id: Ifb625c11b0734a72de08c2dc35b5b85405bbb6f3

opcua-tsn-team-kalycito committed:
[refactor]: Revert the client examples changes

Signed-off-by: Keerthivasan Alagarsamy Senthilkumaran <keerthivasan.as@kalycito.com>
Change-Id: I5a9874705be0b6c4bef4ca922c1a643e78fdd742

opcua-tsn-team-kalycito committed:
[refactor]: Revert the sample application of user database

Signed-off-by: Keerthivasan Alagarsamy Senthilkumaran <keerthivasan.as@kalycito.com>
Change-Id: Id0caf384196a5db534736ee1e8376a689da153f0

opcua-tsn-team-kalycito committed:
[refactor]: Remove unwanted files

Signed-off-by: Keerthivasan Alagarsamy Senthilkumaran <keerthivasan.as@kalycito.com>
Change-Id: I479445f41e6f81ac5825176fe9e4e77c701a0388

opcua-tsn-team-kalycito committed:
[refactor]: Enable conditional checks for role permission

Signed-off-by: Keerthivasan Alagarsamy Senthilkumaran <keerthivasan.as@kalycito.com>
Change-Id: Ib75d6ac5e7f99e6fd6e8a4e6068aaa0ab53e4eb6

AsishGanesh1 committed:
Add sample application for role based access control
 - Addressed following comments from Julius:
   1. Trailing "s" missing for RolePermissions
   2. Include all well known roles
   3. Use prefered naming scheme
   4. Rename AccessControlGroup

Signed-off-by: Asish Ganesh <asish@eclatron.com>

AsishGanesh1 committed:
Attach well known roles to session

 - [X] Add 8 well known role details to session instance
       UA_Server_getSessionAttribute API can be made use of for accessing the role data
 - [X] Remove redundant API's (like hasAccessToNode, hasAccessToMethod)
 - [X] Remove read user AccessLevel/Executable/RolePermissions related methods
       To avoid read outside the context of a session

 - TODO:
 - [ ] Add CMake flag for usage of role based permission
 - [ ] Add test application
 - [ ] Address remaining review comments

Signed-off-by: Asish Ganesh <asish@eclatron.com>

AsishGanesh1 commented:
This pull request is build on top of #5636.
Track it here for further updates.

jpfr commented on src/server/ua_session.h:
```diff
@@ -57,6 +57,9 @@ typedef struct {
     UA_String *localeIds;
 
+    /* Role information */
+    UA_NodeId role;
```
Does a session have only a single role?
I would guess that for fine-grained roles we would like to have several roles per session.

AsishGanesh1 replied:
I've incorporated managing multiple roles per session.

jpfr replied:
Nice.
Since every role gets a power of two 2,4,8,16,32,… we can use a 64Bit integer as bitmap for all the roles. That is more efficient also for comparing inside the nodes..

jpfr commented on CMakeLists.txt:
```diff
@@ -186,7 +186,7 @@ endif()
 set(UA_MULTITHREADING ${MULTITHREADING_DEFAULT} CACHE STRING
     "Multithreading support (<100: No multithreading support, >=100: Thread-safety enabled (internal mutexes)")
 
-option(UA_ENABLE_SUBSCRIPTIONS_ALARMS_CONDITIONS "Enable the use of A&C. (EXPERIMENTAL)" OFF)
```
A&C doesn't have much to do with the roles.
Let's look at the status of A&C in a second step.

jpfr commented on include/open62541/client_highlevel.h:
```diff
@@ -454,6 +464,16 @@ UA_Client_writeUserExecutableAttribute(UA_Client *client, const UA_NodeId nodeId
+                                       size_t newRolePermissionSize,
+                                       const UA_RolePermissionType *rolePermission);
+
+UA_StatusCode UA_EXPORT
```
My guess is that the UserRolePermissions are auto-computed from the roles of the user and the permissions of these roles.
Similar to AccessRights vs UserAccessRights.
So this is not something that can be written, only read.
Check if `writeUserRolePermissionsAttribute` makes sense or should be removed.

AsishGanesh1 replied:
`writeUserRolePermissionsAttribute` is removed

jpfr commented on include/open62541/plugin/accesscontrol.h:
```diff
@@ -80,27 +81,36 @@ struct UA_AccessControl {
     UA_Boolean (*allowAddNode)(UA_Server *server, UA_AccessControl *ac,
                                const UA_NodeId *sessionId, void *sessionContext,
-                               const UA_AddNodesItem *item);
+                               const UA_AddNodesItem *item, UA_RolePermissionType *userRolePermission, size_t userRoleSize);
```
I think we can keep the current API intact.
By adding a lightweight method where we can look up the userRolePermissions from the session id.
So `allowNode` can read the permissions only when needed.

jpfr commented on include/open62541/plugin/accesscontrol.h:
```diff
@@ -80,27 +81,36 @@ struct UA_AccessControl {
-                                  const UA_NodeId *nodeId, void *nodeContext);
+                                  const UA_NodeId *nodeId, void *nodeContext, UA_RolePermissionType *userRolePermission,size_t userRoleSize);
+
+    /* Check access to Node */
```
Add more documentation.
How is `access` different from allowRead and allowBrowse?

AsishGanesh1 replied:
These API's are removed.

jpfr commented on include/open62541/plugin/accesscontrol.h:
```diff
@@ -127,6 +137,8 @@ struct UA_AccessControl {
                                                       UA_DateTime endTimestamp,
                                                       bool isDeleteModified);
 #endif
+    UA_Boolean (*checkUserDatabase)(const UA_UserNameIdentityToken *userToken, UA_String *roleName);
```
What does this do?
Validate that the user has a certain role?

AsishGanesh1 replied:
`checkUserDatabase` is not required and is removed.

jpfr commented on include/open62541/plugin/nodestore.h:
```diff
@@ -411,6 +412,11 @@ struct UA_NodeHead {
                                        * DataChanges (no sampling interval). */
 #endif
+
+    size_t rolePermissionsSize;
```
It seems to be a lot of overhead to manage all role-permissions for each node.
Does the standard define a way how permissions propagate along the hierarchical references?

The best solution would be to inherit standard role-permissions for every new node based on the parents.
And require manual role definitions only as override to the inherited permissions.

jpfr commented on include/open62541/plugin/nodestore.h:
```diff
@@ -411,6 +412,11 @@ struct UA_NodeHead {
+
+    size_t rolePermissionsSize;
+    UA_RolePermissionType *rolePermissions;
+    size_t userRolePermissionsSize;
```
The user-role-permissions are the result of compating the userroles and the role-permissions.
They are different for every user.
So it doesn't make sense to store them in the node for all users at once.

jpfr reviewed (commented):
Nice! Good to see a contribution from your end.
You have to help us a bit getting our head around this part of the spec.
Comments inline.

AsishGanesh1 committed:
Handle several roles per session

 - Remove unused API's & redundant code sections
 - Each session can have a user logged in who can have multiples roles

Signed-off-by: Asish Ganesh <asish@eclatron.com>

AsishGanesh1 commented on include/open62541/server.h:
```diff
@@ -711,6 +713,13 @@ UA_Server_readAccessLevel(UA_Server *server, const UA_NodeId nodeId,
 }
 
+static UA_INLINE UA_THREADSAFE UA_StatusCode
+UA_Server_readUserAccessLevel(UA_Server *server, const UA_NodeId nodeId,
```
This method is removed.

jpfr commented on plugins/ua_accesscontrol_default.c:
```diff
@@ -7,6 +7,7 @@
  */
 
 #include <open62541/plugin/accesscontrol_default.h>
+#include <server/ua_server_internal.h>
```
The server_internal.h is not public for the plugins to use…

jpfr commented on src/server/ua_services.h:
```diff
@@ -50,6 +50,27 @@ typedef void (*UA_Service)(UA_Server*, UA_Session*,
 typedef void (*UA_ChannelService)(UA_Server*, UA_SecureChannel*,
                                   const void *request, void *response);
 
+typedef enum {
```
Where are custom roles assigned to a number?

jpfr reviewed (commented):
Very nice.
I am still missing how the access control plugin can access the roles via the API.
That should be possible as well.

i would like to split this into two PR.
The first adds roles to the sessions and exposes them to accesscontrol.
The second one adds roles to the nodes as well to match.

andreasebner commented:
Thank's for shareing this ideas. A RBAC implementation is part of the latest master.
